platform/pulpcore/app/models/storage.py

```diff
@@ -101,300 +97,86 @@ def __call__(self, model, name):
-                pass
-
-        return path
-
-    @staticmethod
-    def get_artifact_path(sha256digest):
-        """
-        Determine the absolute path where a file backing the Artifact should be stored.
-
-        Args:
-            sha256digest (str): sha256 digest of the file for the Artifact
-
-        Returns:
-            A string representing the absolute path where a file backing the Artifact should be
-            stored
-        """
-        return os.path.join(settings.MEDIA_ROOT, 'artifact', sha256digest[0:2], sha256digest[2:])
+        if name is None:
+            name = content.name
 
-    @staticmethod
-    def published_metadata_path(model, name):
-        """
-        Get the storage path for published metadata.
-
-        Args:
-            model (pulpcore.app.models.PublishedMetadata): A model instance.
-            name (str): The file name.
-
-        Returns:
-            str: The absolute storage path.
-        """
-        return os.path.join(
-            settings.MEDIA_ROOT,
-            'published',
-            'metadata',
-            str(model.pk),
-            name)
-
-    def get_available_name(self, name, max_length=None):
-        """
-        Get the available absolute path based on the name requested.
-        Required by the Storage API.
-
-        Args:
-            name (str): File name.
-            max_length (int): Maximum length, in characters, of the returned path.
-
-        Returns:
-            str: Available name.
-
-        Raises:
-            ValueError if ``max_length`` exceeded.
-        """
-        if max_length:
-            if len(name) > max_length:
-                raise ValueError('max_length exceeded')
-        return name
-
-    def delete(self, path):
-        """
-        Delete the file at the specified path.
-        Required by the Storage API.
+        if not hasattr(content, 'chunks'):
+            content = File(content, name)
 
-        Args:
-            path (str): An absolute file path.
-        """
-        FileSystem.unlink(path)
-
-    def exists(self, path):
-        """
-        Get whether the file actually exists at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute file path.
-
-        Returns:
-            bool: True if exists.
-        """
-        return os.path.exists(path)
-
-    def listdir(self, path):
-        """
-        List the content of the directory at the specified path.
-        Required by the Storage API.
-
-        Args:
-            path (str): An absolute directory path.
-
-        Returns:
-            A tuple of two lists: (directories, files).
-        """
-        files = []
-        directories = []
-        for entry in os.listdir(path):
-            if os.path.isdir(os.path.join(path, entry)):
-                directories.append(entry)
+        try:
+            name = self.get_available_name(name, max_length=max_length)
+            return self._save(name, content)
```

Comment:
Delegating everything FileSystemStorage._save() has reintroduced a lot of bad and unwanted behavior.
- I thought there was a race condition flaw with directory creation logic but after looking again it does catch EEXIST.  So, although the logic is poor, it works.
- Problems with `file_move_safe()`
  - Logic related to *allow_overwrite* has race condition flaw.
  - Uses `os.reneme()` which removes the source (storage should not do this) and has the _nasty_ behavior of silently overwriting the target.
- Determines new filename when already exists.

What you had before with the _save() was better.  Just needed to deal with `path()` and `get_available_name()` overrides.